util/system/path.go

```diff
@@ -193,9 +193,11 @@ func CheckSystemDriveAndRemoveDriveLetter(path string, inputOS string) (string,
 	}
 
 	parts := strings.SplitN(path, ":", 2)
+
+	// we need to preserve the `/` or `\\` at the end because #5249
 	// Path does not have a drive letter. Just return it.
 	if len(parts) < 2 {
-		return ToSlash(filepath.Clean(path), inputOS), nil
+		return ToSlash(filepath.Clean(path), inputOS) + keepTrailingSlash(parts[0], inputOS), nil
```

Comment:
`filepath.Clean` is still current system dependent here.

Reply:
Let me rethink the whole of this part, to a more simplified approach.

Reply:
LMK what you think about this new update. I've made sure that the trailing slash is only kept in the form that it was in the original path `\\` or `/`; and since `filepath.Clean()` is system dependent, it's only appended if it has not been cleaned up in the first place.

On Linux vs. Windows for instance:
```go
p := "\\sample\\"

// on linux
fmt.Printf("%v\n", p == filepath.Clean(p)) // true

// on windows
fmt.Printf("%v\n", p == filepath.Clean(p)) // true
```

Reply:
@tonistiigi -- what do you think about handling this for WCOW scenarios only for now? So I made platform specific implementations of `keepTrailingSlash`.

Reply:
Why is this issue not appearing in linux? Isn't the issue in https://github.com/moby/buildkit/blob/v0.16.0/util/system/path.go#L75-L76 that is currently linux-only and doesn't handle `inputOS` .

Reply:
Thanks for the catch, yes, the issue also silently appears in Linux cross-builds. Silently coz there's no `RUN` to assertain the failure.

Repro dockerfile without `RUN`:
```dockerfile
FROM mcr.microsoft.com/windows/nanoserver:ltsc2022
COPY test1.txt /sample/
#RUN type \sample\test1.txt
```

Cross-build:
```
$ docker buildx build --platform windows/amd64 --builder buildkitd-dev --no-cache --tag=windows-test . --progress plain --output type=local,dest=./output

#6 [2/2] COPY test1.txt /sample/
#6 DONE 0.1s

#7 exporting to client directory
#7 copying files 31B
#7 copying files 230.88MB 0.8s done
#7 DONE 0.8s
```


Checking results, `sample` is a file instead of a directory:
```
$ ls -l output/sample
-rw-r--r-- 1 root root 6 Sep 24 08:57 output/sample
```

Fixing for Linux too.

Reply:
And as for this:
>  Isn't the issue in https://github.com/moby/buildkit/blob/v0.16.0/util/system/path.go#L75-L76 that is currently linux-only and doesn't handle inputOS .

by the time we get to calling `NormalizePath`, `origPath` will already be getting a stripped out `/`, so there's nothing to keep.
```GO
> github.com/moby/buildkit/util/system.NormalizePath() /home/nandaa/dev/oss/buildkit/util/system/path.go:40 (PC: 0x1cfeb6d)
    35: func NormalizePath(parent, newPath, inputOS string, keepSlash bool) (string, error) {
    36:         if inputOS == "" {
    37:                 inputOS = "linux"
    38:         }
    39:
=>  40:         newPath = ToSlash(newPath, inputOS)
    41:         parent = ToSlash(parent, inputOS)
    42:         origPath := newPath
    43:
    44:         if parent == "" {
    45:                 parent = "/"
(dlv) p newPath
"/sample"
```

It is stripped out at `CheckSystemDriveAndRemoveDriveLetter` a few steps before, L1787:
![image](https://github.com/user-attachments/assets/0d30086c-4c8d-4e0a-9de0-3c27998cd74b)

Reply:
So, `NormalizePath` expected that there will be a trailing slash, hence the `keepSlash` bool, but it ends up being stripped away before we call into it.

Reply:
Now fixed for both WCOW and cross-building from Linux in [`diff 36e39..1a528`](https://github.com/moby/buildkit/compare/36e398ea1fcff939f553e239be6cf0cf0eb3289c..1a5289cb21f9b93cb5666bd8696692a384f08143). PTAL.